# Ejercicio de máquinas de vector soporte

Se realizará una práctica para demostrar cómo funcionan las **máquinas de vector soporte** o **SVM** por sus siglas en inglés, haciendo uso de una Jupyter Notebook con Python 3.8 y la librería `scikit-learn`, la cual implementa diversos algoritmos y técnicas de aprendizaje máquina, incluyendo SVM.

El primer paso es cargar el conjunto de datos para clasificarlo mediante el algoritmo de SVM. Para este ejercicio se usarán datos obtenidos de imágenes de cáncer de mama, la cual tiene dos clases de cáncer: maligno o benigno.

In [6]:
from sklearn.datasets import load_breast_cancer

cancer_dataset = load_breast_cancer()
print('Características:', cancer_dataset.feature_names)
print('\nCantidad de características:', len(cancer_dataset.feature_names))
print('\nClases:', cancer_dataset.target_names)

Características: ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']

Cantidad de características: 30

Clases: ['malignant' 'benign']


Ahora vamos a dividir el conjunto de datos actual en un conjunto de entrenamiento, para alimentar al algoritmo, y otro de prueba, para medir qué tan bueno es.

In [7]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(
    cancer_dataset.data,
    cancer_dataset.target,
)

Vamos a importar el módulo de la librería `sk-learn` que nos permitirá aplicar el SVM y construir un objeto de este tipo para usar los métodos correspondientes a este algoritmo.

In [9]:
from sklearn.svm import SVC

classifier = SVC(kernel='linear', gamma='scale')
# entrenar el clasificador de SVM
classifier.fit(x_train, y_train);

Con el modelo ya entrenado podemos pasar a evaluar qué tan preciso es al clasificar datos que no fueron usados en el entrenamiento.

In [11]:
from sklearn import metrics

# predecir datos de prueba con modelo entrenado
y_predicted = classifier.predict(x_test)
# evaluar modelo
metrics.accuracy_score(y_test, y_predicted)

0.9370629370629371

Este resultado nos dice que esta máquina de vector soporte clasifica correctamente los datos el 93.7% de las veces.